File: Largest_rectangular_sub_matrix_whose_sum_is_0.py
class Solution:
  def findSum(self, r1, c1, r2, c2, n, m, pref):
    sum = pref[r2][c2]
    if (c1 - 1) >= 0:
        sum -= pref[r2][c1 - 1]
    if (r1 - 1) >= 0:
        sum -= pref[r1 - 1][c2]
    if (c1 - 1) >= 0 and (r1 - 1) >= 0:
        sum += pref[r1 - 1][c1 - 1]
    return sum
  def sumZeroMatrix(self, a : List[List[int]]) -> List[List[int]]:
    n = len(a)
    m = len(a[0])
    pref = [[0] * m for _ in range(n)]
    for i in range(n):
        for j in range(m):
            pref[i][j] = a[i][j]
            if (j - 1) >= 0:
                pref[i][j] += pref[i][j - 1]
    for j in range(m):
        for i in range(n):
            if (i - 1) >= 0:
                pref[i][j] += pref[i - 1][j]
    ans = 0
    sub_matrix = [0, 0, 0, 0]
    for r1 in range(n):
        for c1 in range(m):
            for r2 in range(r1, n):
                for c2 in range(c1, m):
                    sum = self.findSum(r1, c1, r2, c2, n, m, pref)
                    if sum == 0:
                        submatrixSize = (r2 - r1 + 1) * (c2 - c1 + 1)
                        if submatrixSize > ans:
                            ans = submatrixSize
                            sub_matrix = [r1, r2, c1, c2]
                        elif submatrixSize == ans:
                            if c1 < sub_matrix[2]:
                                sub_matrix = [r1, r2, c1, c2]
                            elif c1 == sub_matrix[2] and r2 - r1 + 1 > sub_matrix[1] - sub_matrix[0] + 1:
                                sub_matrix = [r1, r2, c1, c2]
                            elif c1 == sub_matrix[2] and r2 - r1 + 1 == sub_matrix[1] - sub_matrix[0] + 1 and r1 < sub_matrix[0]:
                                sub_matrix = [r1, r2, c1, c2]
    if ans == 0:
        result = [[]]
        return result
    first_row = sub_matrix[0]
    last_row = sub_matrix[1]
    first_col = sub_matrix[2]
    last_col = sub_matrix[3]
    ans_mat = [[0] * (last_col - first_col + 1) for _ in range(last_row - first_row + 1)]
    for i in range(first_row, last_row + 1):
        for j in range(first_col, last_col + 1):
            ans_mat[i - first_row][j - first_col] = a[i][j]
    return ans_mat

File: Largest_rectangular_sub_matrix_whose_sum_is_0.java
class Solution {
     public static int findSum(int r1, int c1, int r2, int c2, int n, int m, int[][] pref){
        int sum = pref[r2][c2];
        if ((c1 - 1) >= 0)
        {
            sum -= pref[r2][c1 - 1];
        }
        if ((r1 - 1) >= 0)
        {
            sum -= pref[r1 - 1][c2];
        }
        if ((c1 - 1) >= 0 && (r1 - 1) >= 0)
        {
            sum += pref[r1 - 1][c1 - 1];
        }
        return sum;
    }

   
        public static ArrayList<ArrayList<Integer>> sumZeroMatrix(int[][] mat) {
            int n = mat.length;
            int m = mat[0].length;
            
        int[][] pref = new int[n][m];
        for(int i = 0; i<n; i++) 
        {
            for(int j = 0; j<m; j++) 
            {
                pref[i][j] = 0;
            }
        }

        // Row-wise sum.
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                pref[i][j] = mat[i][j];
                if ((j - 1) >= 0)
                {
                    pref[i][j] += pref[i][j - 1];
                }
            }
        }

        // Column-wise sum.
        for (int j = 0; j < m; j++)
        {
            for (int i = 0; i < n; i++)
            {
                if ((i - 1) >= 0)
                {
                    pref[i][j] += pref[i - 1][j];
                }
            }
        }

        int ans = 0;
        int r1=0;
        int r2=0;
        int c1=0;
        int c2=0;
         int[] r1r2c1c2 = new int[4];
        for ( r1 = 0; r1 < n; r1++)
        {
            for ( c1 = 0; c1 < m; c1++)
            {
                for ( r2 = r1; r2 < n; r2++)
                {
                    for ( c2 = c1; c2 < m; c2++)
                    {
                        // Find the sum of submatrix(r1, c1, r2, c2).
                        int sum = findSum(r1, c1, r2, c2, n, m, pref);

                        if (sum == 0)
                        {
                            int submatrixSize = (r2 - r1 + 1) * (c2 - c1 + 1);
                                             if(ans<submatrixSize){
                                ans=submatrixSize;
                            //    System.out.println(r1+" "+r2+"       ");
                            
                                r1r2c1c2=new int[]{r1,r2,c1,c2};
                            }
                            if(ans==submatrixSize){
                                if(c1<r1r2c1c2[2])r1r2c1c2=new int[]{r1,r2,c1,c2};
                            }
                            
                        }
                    }
                }
            }
        }
        int ii=0;
        ArrayList<ArrayList<Integer>> al = new ArrayList<>();
        
        r1=r1r2c1c2[0];
        r2=r1r2c1c2[1];
        c1=r1r2c1c2[2];
        c2=r1r2c1c2[3];
        if(r1==0 && r2==0 && c1==0 && c2==0 && mat[0][0]!=0){
            al.add(new ArrayList<>());
            al.get(0).add(-1);
            return al;
        }
        for(int i = r1 ; i<=r2 ;i++){
            al.add(new ArrayList<>());
            for(int j = c1 ; j<=c2;j++){
                al.get(ii).add(mat[i][j]);
            }
            ii++;
        }
        return al;        
    }
}
